synapse/storage/data_stores/main/state.py

```diff
@@ -1044,6 +1045,12 @@ def __init__(self, db_conn, hs):
             table="event_to_state_groups",
             columns=["state_group"],
         )
+        self.register_background_index_update(
```

Comment:
I'm not thrilled about adding yet another index to postgres which we're going to have to maintain in the event persistence path.

Could we shuffle the logic around so that we can delete the state groups first, and then the events?

Reply:
That would involve loading all the state groups IDs that are to be deleted into memory, which might be large for large rooms? I guess a list of ints isn't actually going to consume too much memory

Reply:
(Also, this doesn't technically happen in the event persistence path, as we store new state groups separately)

Reply:
Done, though this won't delete "orphaned" state groups for a room, but it didn't before so :man_shrugging: